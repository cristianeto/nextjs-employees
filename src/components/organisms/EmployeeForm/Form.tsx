import React from 'react';
import { Button } from '@chakra-ui/react';
import { useFormik } from 'formik';
import employeeSchema from './employeeSchema';
import { ButtonGroup } from './form.styles';
import { TextField } from '@atoms';
import { employeeForm } from '@constants';
import { IForm } from '@interfaces';

const Form: React.FC<IForm> = ({
  initialState: state,
  onSubmit: submit,
  onClose,
  labels,
}) => {
  const { dni, name: firstname, lastname, email } = labels;
  const {
    titles: { saveRegister },
  } = employeeForm;

  const formik: any = useFormik({
    initialValues: state,
    onSubmit: (values) => {
      submit(values);
    },
    validationSchema: employeeSchema,
  });

  const register = (name: string) => ({
    isError: formik.errors[name] && formik.touched[name],
    helperText: formik.errors[name],
    name,
    onChange: formik.handleChange,
    value: formik.values[name],
    onBlur: formik.handleBlur,
  });

  return (
    <form onSubmit={formik.handleSubmit}>
      <TextField {...register('dni')} label={dni} placeholder="DNI" />
      <TextField {...register('name')} label={firstname} placeholder="Name" />
      <TextField
        {...register('lastname')}
        label={lastname}
        placeholder="Lastname"
      />
      <TextField
        {...register('email')}
        label={email}
        placeholder="Email"
        type="email"
      />
      <ButtonGroup gap={4} width="100%">
        <Button colorScheme="blue" type="submit">
          {saveRegister}
        </Button>
        <Button variant="outline" onClick={onClose}>
          Cancel
        </Button>
      </ButtonGroup>
    </form>
  );
};

export default Form;
