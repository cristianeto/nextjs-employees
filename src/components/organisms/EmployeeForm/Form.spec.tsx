import { render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Form from './Form';
import { employeeForm, initialState, employee } from '@mocks';

describe('<Form/>', () => {
  it('should render the component', () => {
    const { baseElement } = render(
      <Form
        initialState={initialState}
        labels={employeeForm.labels}
        onClose={jest.fn()}
        onSubmit={jest.fn()}
      />,
    );
    expect(baseElement).toBeTruthy();
  });

  it('should render save button', () => {
    const onSubmit = jest.fn();
    const { getByRole } = render(
      <Form
        initialState={initialState}
        labels={employeeForm.labels}
        onClose={jest.fn()}
        onSubmit={onSubmit}
      />,
    );
    const button = getByRole('button', {
      name: `${employeeForm.titles.saveRegister}`,
    });
    expect(button).toBeTruthy();
    expect(button).toBeInTheDocument();
  });

  it('should allow user type inputs', async () => {
    const onSubmit = jest.fn();
    const { getByRole } = render(
      <Form
        initialState={initialState}
        labels={employeeForm.labels}
        onClose={jest.fn()}
        onSubmit={onSubmit}
      />,
    );
    const dniInput = getByRole('textbox', { name: /dni/i });
    const nameInput = getByRole('textbox', { name: 'Name' });
    const lastnameInput = getByRole('textbox', { name: /lastname/i });
    const emailInput = getByRole('textbox', { name: /email/i });
    expect(dniInput).toBeTruthy();
    await userEvent.type(dniInput, employee.dni);
    await userEvent.type(nameInput, employee.name);
    await userEvent.type(lastnameInput, employee.lastname);
    await userEvent.type(emailInput, employee.email);
    expect(dniInput).toHaveValue(employee.dni);
    expect(nameInput).toHaveValue(employee.name);
    expect(lastnameInput).toHaveValue(employee.lastname);
    expect(emailInput).toHaveValue(employee.email);
  });

  it('should allow user submit the form', async () => {
    const handleSubmit = jest.fn();
    const { getByRole } = render(
      <Form
        initialState={initialState}
        labels={employeeForm.labels}
        onClose={jest.fn()}
        onSubmit={handleSubmit}
      />,
    );
    const user = userEvent.setup();
    const dniInput = getByRole('textbox', { name: /dni/i });
    const nameInput = getByRole('textbox', { name: 'Name' });
    const lastnameInput = getByRole('textbox', { name: /lastname/i });
    const emailInput = getByRole('textbox', { name: /email/i });
    expect(dniInput).toBeTruthy();
    await user.type(dniInput, employee.dni);
    await user.type(nameInput, employee.name);
    await user.type(lastnameInput, employee.lastname);
    await user.type(emailInput, employee.email);
    const submitButton = getByRole('button', { name: /save/i });
    await user.click(submitButton);
    await waitFor(() => expect(handleSubmit).toHaveBeenCalledTimes(1));
  });
});
