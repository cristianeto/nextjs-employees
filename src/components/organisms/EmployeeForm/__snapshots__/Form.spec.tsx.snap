// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form/> should render the component 1`] = `
<body>
  <div>
    <form>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-0"
          for="field-:r0:"
          id="field-:r0:-label"
        >
          Dni
        </label>
        <input
          class="chakra-input css-0"
          id="field-:r0:"
          name="dni"
          placeholder="DNI"
          type="text"
          value=""
        />
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-0"
          for="field-:r1:"
          id="field-:r1:-label"
        >
          Name
        </label>
        <input
          class="chakra-input css-0"
          id="field-:r1:"
          name="name"
          placeholder="Name"
          type="text"
          value=""
        />
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-0"
          for="field-:r2:"
          id="field-:r2:-label"
        >
          Lastname
        </label>
        <input
          class="chakra-input css-0"
          id="field-:r2:"
          name="lastname"
          placeholder="Lastname"
          type="text"
          value=""
        />
      </div>
      <div
        class="chakra-form-control css-0"
        role="group"
      >
        <label
          class="chakra-form__label css-0"
          for="field-:r3:"
          id="field-:r3:-label"
        >
          Email
        </label>
        <input
          class="chakra-input css-0"
          id="field-:r3:"
          name="email"
          placeholder="Email"
          type="email"
          value=""
        />
      </div>
      <div
        class="chakra-button__group css-1yped2f"
        role="group"
      >
        <button
          class="chakra-button css-0"
          type="submit"
        >
          Save
        </button>
        <button
          class="chakra-button css-0"
          type="button"
        >
          Cancel
        </button>
      </div>
    </form>
  </div>
</body>
`;
